### PR TITLE
Enabling sort optimization back for half_float with custom comparators

### PR DESCRIPTION
**Pull request by gashutos**
### Description
Enabling back sort optimization for `half_float` which was disabled here https://github.com/opensearch-project/OpenSearch/pull/10999.
As mentioned in below issue, for this sort optimization to get enabled, the BKD indexed point size and point size passed in comparator should be exactly same. With custom comparator `HalfFloatComparator` we achieved that.
Also I have added missing unit tests & integ tests for `half_float` field type to avoid any such issues.

[I am planning to add this in opensearch-3.0]

### Tests
1. Added `half_float` field type sort unit tests.
2. Added `half_float` field type sort integ tests.
3. Added `half_float` field type `search_after` tests.
4. Added `half_float` field type mixed type sort result tests.
5. Added `half_float` field type missign value sort tests.
6. Added `half_float` field type as index sort.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
#10998

### Perf impact with this change (this is already there but we had to disable)
Before this change
```
|                                   50th percentile service time | desc_sort_tip_amount |     1478.62 |     ms |
|                                   90th percentile service time | desc_sort_tip_amount |     1484.97 |     ms |
|                                   99th percentile service time | desc_sort_tip_amount |     1487.94 |     ms |
|                                  100th percentile service time | desc_sort_tip_amount |     1499.31 |     ms |

|                                   50th percentile service time |  asc_sort_tip_amount |     1466.53 |     ms |
|                                   90th percentile service time |  asc_sort_tip_amount |     1473.64 |     ms |
|                                   99th percentile service time |  asc_sort_tip_amount |     1496.76 |     ms |
|                                  100th percentile service time |  asc_sort_tip_amount |     1704.93 |     ms |
```
After this change
```
|                                   50th percentile service time | desc_sort_tip_amount |     9.26986 |     ms |
|                                   90th percentile service time | desc_sort_tip_amount |     9.82747 |     ms |
|                                   99th percentile service time | desc_sort_tip_amount |     10.3689 |     ms |
|                                  100th percentile service time | desc_sort_tip_amount |     10.4408 |     ms |

|                                   50th percentile service time |  asc_sort_tip_amount |     95.3346 |     ms |
|                                   90th percentile service time |  asc_sort_tip_amount |      95.776 |     ms |
|                                   99th percentile service time |  asc_sort_tip_amount |     100.285 |     ms |
|                                  100th percentile service time |  asc_sort_tip_amount |       109.3 |     ms |
```
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gashutos** [comment]
@reta @backslasht I have added back half_float optimization with custom comparators introductions. Added much more tests for half_float now.

File: server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /** Leaf comparator for {@link HalfFloatComparator} that provides skipping functionality */
+    public class HalfFloatLeafComparator extends NumericLeafComparator {
+
+        public HalfFloatLeafComparator(LeafReaderContext context) throws IOException {
+            super(context);
+        }
+
+        private float getValueForDoc(int doc) throws IOException {
+            if (docValues.advanceExact(doc)) {
+                return Float.intBitsToFloat((int) docValues.longValue());
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java:63]: Have some doubts about this conversion: when stored in doc values, the value is encoded using `HalfFloatPoint.halfFloatToSortableShort`, I am wondering if we should decode it using `HalfFloatPoint.sortableShortToHalfFloat` ?

**Comment by reta** [comment]
Could we use half float for index sort? Could you please add test case to `indices.sort/10_basic.yml` please?

File: server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.sandbox.document.HalfFloatPoint;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.comparators.NumericComparator;
+
+import java.io.IOException;
+
+/**
+ * The comparator for unsigned long numeric type.
```
[the remark this answers is not in the document]
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java]: Done

**Comment by gashutos** [comment]
> Could we use half float for index sort? Could you please add test case to `indices.sort/10_basic.yml` please?

Yes ! i have checked this, 
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/IndexSortConfig.java#L268C5-L274C7

`half_float` field is not allowed for index sort field.

**Comment by reta** [comment]
> `half_float` field is not allowed for index sort field.

Would be great to have tests (that is it not allowed) and update the documentation for it (I think we have no details at all)

**Comment by reta** [comment]
What about aggregations? It seems live we have zero tests for that ... in rest-api-specs

**Comment by gashutos** [comment]
> > `half_float` field is not allowed for index sort field.
> 
> Would be great to have tests (that is it not allowed) and update the documentation for it (I think we have no details at all)

Yeah makes sense. Added

**Comment by gashutos** [comment]
> What about aggregations? It seems like we have zero tests for that ... in rest-api-specs

This will be bigger change and initiative. Can we do this outside of this PR ? Let me open up an issue on this.

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java [hunk 3]
```diff
@@ -78,7 +78,7 @@ public SortField.Type reducedType() {
         return SortField.Type.FLOAT;
     }
 
-    private NumericDoubleValues getNumericDocValues(LeafReaderContext context, float missingValue) throws IOException {
+    protected NumericDoubleValues getNumericDocValues(LeafReaderContext context, float missingValue) throws IOException {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java:81]: Do you think we should better off copy this method to `HalfFloatXxx` counterpart instead of opening it up?

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java [hunk 4]
```diff
@@ -61,7 +61,7 @@
  */
 public class FloatValuesComparatorSource extends IndexFieldData.XFieldComparatorSource {
 
-    private final IndexNumericFieldData indexFieldData;
+    protected final IndexNumericFieldData indexFieldData;
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java:64]: Same here, let keep the `private` but add package protected `IndexNumericFieldData getIndexFieldData()` method to limit the API exposure.

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java (hunk 3 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java]: +1 , that would be better, done.

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java (hunk 4 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/FloatValuesComparatorSource.java]: makes sense

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
Thanks @gashutos , @msfroh do you want to take a look? thank you!

File: server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.sandbox.document.HalfFloatPoint;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.comparators.NumericComparator;
+
+import java.io.IOException;
+
+/**
+ * The comparator for unsigned half_float numeric type.
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java:19]: I think those are signed, right?

```suggestion
 * The comparator for half_float numeric type.
```

**Comment by gashutos** [comment]
I have added OSB half_float queries as well for sorting to improve benchmark coverage
https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/139

**Comment by gashutos** [comment]
@msfroh can you check if changes look good ?

**Comment by gashutos** [comment]
> I have added OSB half_float queries as well for sorting to improve benchmark coverage [opensearch-project/opensearch-benchmark-workloads#139](https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/139)

I have also updated benchmark numbers with this optimization, which we lost while disabling the half_float type.
This is giving real good numbers. 

Before this change
```
|                                   50th percentile service time | desc_sort_tip_amount |     1478.62 |     ms |
|                                   90th percentile service time | desc_sort_tip_amount |     1484.97 |     ms |
|                                   99th percentile service time | desc_sort_tip_amount |     1487.94 |     ms |
|                                  100th percentile service time | desc_sort_tip_amount |     1499.31 |     ms |

|                                   50th percentile service time |  asc_sort_tip_amount |     1466.53 |     ms |
|                                   90th percentile service time |  asc_sort_tip_amount |     1473.64 |     ms |
|                                   99th percentile service time |  asc_sort_tip_amount |     1496.76 |     ms |
|                                  100th percentile service time |  asc_sort_tip_amount |     1704.93 |     ms |
```
After this change
```
|                                   50th percentile service time | desc_sort_tip_amount |     9.26986 |     ms |
|                                   90th percentile service time | desc_sort_tip_amount |     9.82747 |     ms |
|                                   99th percentile service time | desc_sort_tip_amount |     10.3689 |     ms |
|                                  100th percentile service time | desc_sort_tip_amount |     10.4408 |     ms |

|                                   50th percentile service time |  asc_sort_tip_amount |     95.3346 |     ms |
|                                   90th percentile service time |  asc_sort_tip_amount |      95.776 |     ms |
|                                   99th percentile service time |  asc_sort_tip_amount |     100.285 |     ms |
|                                  100th percentile service time |  asc_sort_tip_amount |       109.3 |     ms |
```

File: server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * Comparator based on {@link Float#compare} for {@code numHits}. This comparator provides a
+ * skipping functionality – an iterator that can skip over non-competitive documents.
+ */
+public class HalfFloatComparator extends NumericComparator<Float> {
+    private final float[] values;
+    protected float topValue;
+    protected float bottom;
+
+    public HalfFloatComparator(int numHits, String field, Float missingValue, boolean reverse, boolean enableSkipping) {
+        super(field, missingValue != null ? missingValue : 0.0f, reverse, enableSkipping, HalfFloatPoint.BYTES);
```
Comment (backslasht) [inline server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java:29]: I assume half float support doesn't support negative values? is that right?

**Review by backslasht — CHANGES_REQUESTED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml [hunk 7]
```diff
@@ -119,3 +120,85 @@
   - match: { status: 400 }
   - match: { error.type: search_phase_execution_exception }
   - match: { error.caused_by.reason: "Can't do sort across indices, as a field has [unsigned_long] type in one index, and different type in another index!" }
+
+---
+"search across indices with mixed long and double and float numeric types":
+  - skip:
+      version: " - 2.10.99"
+      reason: half float was broken before 2.11
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: 2.11 we are releasing a patch. The fix was part of that patch.

[the remark this answers is not in the document]
Comment (backslasht) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: Technically, 2.11.0 is already released, I'm not sure if this meets the criteria to go as patch in 2.11.1 as it is not critical (disabling half_float optimization is, which is already checked in).

[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: https://opensearch.slack.com/archives/C0561HRK961/p1698428388137519

Not for this commit `disabling half_float optimization is, which is already checked in` but it is going with another critical commits for which we are releasing patch. It is fine to include in that case for 2.11.1 right ?

[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: It already got in though.
https://github.com/opensearch-project/OpenSearch/pull/11005

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml [hunk 8]
```diff
@@ -119,3 +120,85 @@
...
+        body:
+          sort: [{ counter: desc }]
+  - match: { hits.total: 5 }
+  - length: { hits.hits: 5 }
+  - match: { hits.hits.0._index: test_2 }
+  - match: { hits.hits.0._source.counter: 1223372036854775800.23 }
+  - match: { hits.hits.0.sort.0: 1223372036854775800.23 }
+  - match: { hits.hits.1._index: test_1 }
+  - match: { hits.hits.1._source.counter: 223372036854775800 }
+  - match: { hits.hits.1.sort.0: 223372036854775800  }
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:190]: Done

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml [hunk 9]
```diff
@@ -320,3 +320,75 @@
   - length: { hits.hits: 1 }
   - match: { hits.hits.0._index: test }
   - match: { hits.hits.0._source.population: null }
+
+---
+"half float":
+  - skip:
+        version: " - 2.10.99"
+        reason: half_float was broken for 2.10 and earlier
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:328]: Lets keep it 2.10, as it is part of 2.11.1 patch release.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml [hunk 10]
```diff
@@ -320,3 +320,71 @@
...
+          mappings:
+            properties:
+              population:
+                type: half_float
+  - do:
+      bulk:
+        refresh: true
+        index: test
+        body: |
+          {"index":{}}
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:343]: done

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (backslasht) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: Shouldn't this be 2.11.0 where it is still broken?

[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: For this test framework, it always creates opensearch test cluster from latest patched version, so from 2.11 version family, it will be always 2.11.1. That is the reason this test is passing....

[the remark this answers is not in the document]
Comment (reta) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: @gashutos I think we won't make it to 2.11.1 but 2.12

[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: @reta @backslasht `half_float` type is supported with PR https://github.com/opensearch-project/OpenSearch/pull/10999. Which is included in 2.11.1 patch. With that, the tests writtten here should pass for 2.11. 
This PR we can make it 2.12 which is the optimizing `half_float` sort. So we should be good here for specifying. 2.10.99 istead 2.11 as the test is expected to pass in version 2.11

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:328]: https://github.com/opensearch-project/OpenSearch/pull/11024#discussion_r1382636919
This shuld work fine with 2.11

File: server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java (hunk 6 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/search/comparators/HalfFloatComparator.java:29]: No it has negative values, however we always pass missing value. I added in this new commit, it will never be null for `half_float`. I copoed this from FloatComparator, where it has the same constructor logic.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (reta) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/260_sort_mixed.yml:128]: Got it, thanks for clarifying @gashutos

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml [hunk 11]
```diff
@@ -320,3 +320,130 @@
...
+
+  # search_after with the desc sort
+  - do:
+      search:
+        index: test
+        rest_total_hits_as_int: true
+        body:
+          size: 1
+          sort: [ { population: desc } ]
+          search_after: [ 164.375 ] # this is rounded sort value in sort result
```
Comment (backslasht) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:408]: This rounding is not fully explained here, can you attach the link to the documentation. If in case documentation doesn't exists, can you create a documentation request for the same?

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java [hunk 12]
```diff
@@ -104,7 +104,7 @@ public FieldComparator<?> newComparator(String fieldname, int numHits, boolean e
         final double dMissingValue = (Double) missingObject(missingValue, reversed);
         // NOTE: it's important to pass null as a missing value in the constructor so that
         // the comparator doesn't check docsWithField since we replace missing values in select()
-        return new DoubleComparator(numHits, fieldname, null, reversed, enableSkipping && this.enableSkipping) {
+        return new DoubleComparator(numHits, fieldname, dMissingValue, reversed, enableSkipping && this.enableSkipping) {
```
Comment (backslasht) [inline server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java:107]: is this change intended? If yes, can you add tests for the same?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml [hunk 13]
```diff
@@ -320,3 +320,130 @@
...
+
+  # search_after with the asc sort
+  - do:
+      search:
+        index: test
+        rest_total_hits_as_int: true
+        body:
+          size: 1
+          sort: [ { population: asc } ]
+          search_after: [ 184.375 ] # this is ruonded sort value in sort result
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml]: Lets keep this in  parity with `desc` order. Also `sort` values are recommended to use for `search_after

[the remark this answers is not in the document]
Comment (backslasht) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml]: By sort values do you mean the value returned in the result?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml (hunk 11 above)
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:408]: I have asked question to community in case such documentation exists.
https://github.com/opensearch-project/OpenSearch/issues/11128

There is a precision talk here, https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html#_which_type_should_i_use
But this doesnt say sort query behaviour here like we are seeing. 

Lets take this seperately apart from this PR ?
I will follow up here https://github.com/opensearch-project/OpenSearch/issues/11128 and if we dont find document/answer, I will create doc request.

I myself was not able to find any such documentation though.

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java (hunk 12 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java:107]: Yeah this is intended, adding tests.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (backslasht) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml]: Makes sense! thanks!

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java (hunk 12 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java:107]: I added big test for this, https://github.com/opensearch-project/OpenSearch/pull/11024/commits/a2015ef1c630b169a711e5487312acd3363f608a
For all numeric types. 
Basically, passing missing value is requred in case we have missing values determined, which is the case here.
If we dont pass missing value here, it might hit the corner case like here https://github.com/opensearch-project/OpenSearch/issues/9537. We fixed this in Lucene, https://github.com/apache/lucene/pull/12520, but didnt add integ tests here in OpenSearch. 

Thanks @backslasht for reminding about adding integ tests here. This is definitely one of misssing corner case but important one.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml [hunk 14]
```diff
@@ -320,3 +320,426 @@
...
+  - match: { hits.hits.2._source.int: null }
+  - match: { hits.hits.2._source.id: 22 }
+
+  - do:
+      search:
+        index: test
+        rest_total_hits_as_int: true
+        body:
+          "size": 3
+          "sort": [ { "double": { "order": "asc" } }, { "id": { "order": "asc" } } ]
```
[the remark this answers is not in the document]
Comment (gashutos) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/90_search_after.yml:711]: Oh :(, added.

**Review by backslasht — APPROVED** [review]
Thanks @gashutos for making this changes!

**Review by kkmr — APPROVED** [review]

**Comment by gashutos** [comment]
Thanks @backslasht & @reta for the review !
@reta @msfroh lets merge it if you guys are fine as well ?

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java [hunk 15]
```diff
@@ -104,7 +104,7 @@ public FieldComparator<?> newComparator(String fieldname, int numHits, boolean e
         final double dMissingValue = (Double) missingObject(missingValue, reversed);
         // NOTE: it's important to pass null as a missing value in the constructor so that
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java:105]: @gashutos I think the comment does not match the code anymore but it was here for a reason

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java [hunk 16]
```diff
@@ -120,7 +120,7 @@ public FieldComparator<?> newComparator(String fieldname, int numHits, boolean e
         final long lMissingValue = (Long) missingObject(missingValue, reversed);
         // NOTE: it's important to pass null as a missing value in the constructor so that
         // the comparator doesn't check docsWithField since we replace missing values in select()
-        return new LongComparator(numHits, fieldname, null, reversed, enableSkipping && this.enableSkipping) {
+        return new LongComparator(numHits, fieldname, lMissingValue, reversed, enableSkipping && this.enableSkipping) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java:123]: The change of missingValue argument seems to be unrelated to the subject of this pull request, I feel strongly we should move it off to separate pull request

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java (hunk 15 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java:105]: Yeah, this looke very old comments. This was part of this commit https://github.com/opensearch-project/OpenSearch/commit/629f91ae57b5a2223f2edaa213b5d0d08a155885. which is in 2018 but the skipping logic got introduced later.
For skipping logic, if missing values are competitive, we dont invoke skipping optimization as it skipps missing value docuemts through BKD which were not indexed in BKD at all. https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/comparators/NumericComparator.java#L192C1-L193C1

I added integ tests as well to address this scenario.

I am thinking to remove this comment.

File: server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java (hunk 16 above)
Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java:123]: Sure, let me remove this.

Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java]: Cool, I willl raise PR for missing value once this is merged

Comment (gashutos) [reply server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java]: cc @backslasht @kkmr -> I will be sending new PR for last commit as suggested by reta.
2a4a43009bab047d59aeff65e4166c57bf1796e9

**Comment by dblock** [comment]
This tests seem broken in opensearch-py, https://github.com/opensearch-project/opensearch-py/actions/runs/6868286935/job/18678495702.

**Comment by reta** [comment]
> This tests seem broken when run from opensearch-py, https://github.com/opensearch-project/opensearch-py/actions/runs/6868286935/job/18678495702, [opensearch-project/opensearch-py#582](https://github.com/opensearch-project/opensearch-py/issues/582).

Thanks a lot @dblock , @gashutos mind please taking a look? thank you!

**Comment by dblock** [comment]
I'm also a bit worried seeing that the error message has a serialized `FloatValuesComparatorSource` object in it.

```
{"type":"illegal_argument_exception","reason":"Field population is indexed with 2 bytes per dimension, but org.opensearch.index.fielddata.fieldcomparator.FloatValuesComparatorSource$1@2570133f expected 4"}}
```

**Comment by gashutos** [comment]
@dblock @reta just saw this, it looks like this was invoked in 2.11.0 version where half_float were already broken. The fix is in 2.11.1. The half_float sort test should not have run for 2.11.0. I will check why it did run the test for 2.11.0 for py client. For Ooensearch, this test run only for 2.11.1.

**Comment by dblock** [comment]
> @dblock @reta just saw this, it looks like this was invoked in 2.11.0 version where half_float were already broken. The fix is in 2.11.1. The half_float sort test should not have run for 2.11.0. I will check why it did run the test for 2.11.0 for py client. For OpenSearch, this test run only for 2.11.1.

The REST spec should skip this test for 2.11.0. That version has been released. Will you please make that change @gashutos?

**Comment by dblock** [comment]
> I'm also a bit worried seeing that the error message has a serialized `FloatValuesComparatorSource` object in it.
> 
> ```
> {"type":"illegal_argument_exception","reason":"Field population is indexed with 2 bytes per dimension, but org.opensearch.index.fielddata.fieldcomparator.FloatValuesComparatorSource$1@2570133f expected 4"}}
> ```

@gashutos Is this a new bug?
